### DISCUSSION TITLE
[Bug]: Improved Display Compatibility for Heltec Wireless Stick V3 (64x32) - Bluetooth Code Display

### DISCUSSION THREAD

**Discussion by JanisPlayer** [meshtastic/firmware#3995] [Ideas]
### A notice:
I bought this Heltec from [MoonMall Store](https://de.aliexpress.com/item/1005005212873967.html), I thought it was an official reseller.
In any case, even the documentation for the Heltec Wireless Stick V3 states that it has a 128x64 pixel display.
So I think it's not Heltec's fault, but the shop's fault. I think I was sold an imitation, which you can find on Amazon with a 64x32 pixel display.
By the way, the screen really does slip, as the Amazon reviews say for products that have a 64x32 pixel display.
I have now contacted the dealer and manufacturer and hope they can tell me whether the Heltec Wireless Stick V3 was really original.
I'm thinking about sending it back, so I don't know if I'll take the time to adapt it to such small displays.
I will close the issue if it turns out that I received the wrong product here.
I'm leaving the warning here so that people who encounter the same problem know what they're dealing with.
I'll make an update when I know more and the dealer and manufacturer have responded.
But from the product photos I now know that there are several versions, so be careful so that you get the right one.
Despite the description of 128x64, I got a 64x32 pixel display. (Edit: I recreated what was shown in the picture, same dealer, looks the same, except that my display is a little further down, so I have no idea, let's see what the manufacturer says or maybe you have that too and you know more. This is supposed to be mine according to the Heltec search and it says 128x64 pixels again. [Wireless_Stick_V3/HTIT-WS_V3(Rev1.0).pdf](https://resource.heltec.cn/download/Wireless_Stick_V3/HTIT-WS_V3(Rev1.0).pdf))
So since it's probably not really worth complaining about the display in China, I might even keep it here if the manufacturer says that there are no other differences. I can of course advise against buying it, which is why a lot of the example code doesn't run. By the way, if your display moves, the flex cable will hold it. Otherwise it was loosely attached to me.
Whether the board can work with Mashtastic functions, yes, probably.
But you have to make an IF query whether the display is smaller than 64 in height, if so then the corresponding code must be executed.
What you should perhaps also add would be smaller fronts, these exist, they don't look as nice, but they are better than the text constantly moving.
In the programming example code I strangely found the resolution 64x32, so it's either Heltec's fault or the owners' fault or there are many versions. In any case, as a customer I would have been happy if Heltec or the dealers had provided the information correctly. But as I said, I can only assume here. Sorry that the note was so detailed, but maybe it will help someone who has the same problem.

### Category

BLE

### Hardware

Heltec V3

### Firmware Version

2.2.25.9d37a8d1

### Description

I noticed that the Heltec wireless-stick-v3 has a display error. This seems to be because the information on the manufacturer's website is incorrect. The page: https://heltec.org/project/wireless-stick-v3/ states that it has a 128x64 pixel display, but it is actually a 64x32 pixel display.

I made some changes to https://github.com/meshtastic/firmware/blob/master/src/graphics/Screen.cpp, and at least I was able to display the Bluetooth code. Are there any plans for adapting to such small screens on your end? I am aware that this would involve displaying texts on small screens in a special way, such as moving texts, and programming the recognition of these displays. I have adapted the code on my end so that, for now, only the Bluetooth code is displayed correctly. Such an adjustment is possible with the current code without as many changes as I made.

My test changes:

```cpp
static void drawFrameBluetooth(OLEDDisplay *display, OLEDDisplayUiState *state, int16_t x, int16_t y)
{
    int x_offset = display->width() / 2;
    int y_offset = display->height() <= 32 ? 0 : 16;
    display->setTextAlignment(TEXT_ALIGN_CENTER);
    display->setFont(FONT_SMALL); 

    display->drawString(x_offset + x, y_offset + y, "Bluetooth");

    y_offset += FONT_HEIGHT_SMALL;

    display->drawString(x_offset + x, y_offset + y, "Enter this code");

    String displayPin(btPIN);
    String pin = displayPin.substring(0, 3) + " " + displayPin.substring(3, 6);
    y_offset += FONT_HEIGHT_SMALL - 5;

    display->drawString(x_offset + x, y_offset + y, pin);

    String deviceName = "ID: ";
    deviceName.concat(getDeviceName());
    y_offset += FONT_HEIGHT_SMALL - 5;

    display->drawString(x_offset + x, y_offset + y, deviceName);
}
```

Issue: https://meshtastic.discourse.group/t/heltec-wireless-stick-v3-with-0-49-oled/9158/7

Screenshots:
![Oled64x32](https://github.com/meshtastic/firmware/assets/54918417/e121ba46-15a8-42c0-970e-cbe9cf652ca3)

**Comment by caveman99** [comment]
I have one of these incoming and will take a stab at it. Thanks for the initial work. No promises though :-)

**Comment by JanisPlayer** [comment]
@caveman99  Thank you for the information; now I know that it's probably normal with the resolution. It seems like the product description and documentation might be incorrect.

That's cool! If it works out and they manage to do it, I'll be excited. I've found alternative fonts that can further reduce the text size. With such a small display resolution, displaying only the essentials is necessary, and dealing with long texts requires shortening or making the text move. It's indeed a complex task; I wish you much success. But I believe as long as the Bluetooth code is displayed, it's already sufficient. :)

**Comment by DrBlackross** [comment]
putting battery life, node count and or signal strength on other screen that are toggle-able with the usr button would be cool

(trying to find screen settings in the rest of the code is a bit tricky so far)

**Comment by kiam001** [comment]
It seems that https://heltec.org/project/wireless-stick-v3/ has correct the incorrect information. Additionally, the FAQ states:
Meshtastic firmware for Wireless Stick
Due to the small screen on the board, it currently does not support Meshtastic’s screen display. You can directly download the firmware for the Wireless Stick Lite.

I have now ordered 3 sticks via Amazon and hope that I can use them without the display.

**Reply by JanisPlayer** [reply]
Yes, that's possible, just like with those without a display. You could also customize everything. But that would mean a lot of code. In my case, the seller was accommodating and sent me a LoRa V3 without having to return the old one. So I got one with the correct display. You can then control the other one via commands on the serial interface and set the Bluetooth pin. Displaying messages correctly and so on doesn't really work with the display. But if you only use your smartphone and the Meshtastic API it shouldn't be that bad. https://meshtastic.org/docs/configuration/radio/bluetooth/ The web interface is a bit more convenient if you have a Chrome browser. Then you don't have to enter any commands. https://client.meshtastic.org/

**Comment by kiam001** [comment]
Hope Display become Support in Feature

**Comment by DrBlackross** [comment]
If you flash with the chrome browser, then switch over to the serial
monitor, the Bluetooth pin will (should) be shown there... only used mine
for meshtastic for about a day before flashing it over to tinygs. When the
lora32 came in with the right screen size, then yea it worked, downside
they sent me a V2.1 lora32, soooo, lol have to use the flasher.pxdlocs.com
site to flash for the v2.1 hardware since meshtastics site doesn't have the
firmware for it that can find (and I'm too lazy to compile it).

Just make sure if you order a lora32 (not the wireless stick) it's a V3 or
newer hardware

**Comment by kiam001** [comment]
My works perfekt use the Wireless light Stick Firmware so Display is dark and Pin ist 123456 and runs 4 days without issuses.

Today shortime i try the heltec v3 Firmware Display Turn on but only the Pixel thats have and the Rest was Hidden, i am Back to the Wireless v3 lite Firmware.

In Feature whan the Display is Support i am Happy but device is without Display too poerfull i pay 19€ on amazon for it and the Display Sticks cost current 24€.

This are my First meshtastic try and alrady have on Channel 3 my Home Assistent conversation Assist via local mqtt to running.

The Default Heartbeat lamp i Turn Off and Set the external notify Led on gpio 35 whats the Led is so it Blinks now 10 sec for new Messages

**Comment by miky2k** [comment]
I hope some improvements, because is very cheap version of heltec V3.

**Comment by Lambo52626** [comment]
I wont lie, I wish I saw this post sooner, I struggled with that damn screen, HOWEVER Using the Web portal, I could change the Bluetooth to Fixed and set my pin. but yes what a mission.

FYI Im still wanting to try this,
